crates/sparrow-compiler/src/types/inference.rs

```diff
@@ -25,6 +25,15 @@ pub(crate) fn instantiate(
         "Arguments being instantiated should have been resolved against the signature"
     );
 
+    let mut constraints_for_type_var: HashMap<TypeVariable, Vec<TypeConstraint>> = HashMap::new();
```

Comment:
I'm not sure we need a mutable list of constraints for each type variable. Isn't that information directly avaliable in the type parameters? Similarly, I'm not sure why we still collect types for each type *constraint*.

What I'd expect this code to do:

1. Collect argument types used with each type variable. So `HashMap<TypeVariable, Vec<&Located<FenlType>>>`.
2. For each type variable, determine the concrete type that is the least-upper-bound of those types. So `HashMap<TypeVariable, FenlType>`.
3. For each type variable, make sure that solution satisfies the constraints. This can be done by iterating over the type parameters (since every variable should appear there) and checking the corresponding type against the constraint.

So, we need neither of the HashMaps that are here.